# Master list of programs

**Purpose:** This document is a list all programs/tools on the server and the PATH to the executable. It is also annotated to include special directions for running a program if need be.

**Tip:** If one is going to be using one of these programs regularly, it is recommended one edits one's `.bash_profile` file in one's home directory to export it to PATH every time one logs in. [Instructions on how to do so](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/setting_PATH.md)

**Tip:** If one installs a tool into `/scratch/Tools/`, it is helpful to run `chmod -R +777` on the directory or `chmod +777` on the executable. This ensures that other users have the ability to make changes such as adding packages (e.g., in R or conda), updating the version, or adding/updating databases. 

**Tip:** Always check to see if a program is available as a [conda package](https://anaconda.org/) before installing it locally. If a conda package is available, please create a new environment and install it in to that environment, following the directions [here](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/using_conda.md). Once you have done so, please run `chmod -R +777 /scratch/Tools/anaconda3` to ensure that the next person who wants to install a conda package can.

# /scratch/Tools

This is a list of programs currently located in `/scratch/Tools/`. If you install a program there, please update this list (or contact someone who can edit this page to update it). I have also included links to software homepages, tutorials written by me, and tutorials that I have used written by others. Feel free to add/update tutorials.

- anaconda3
	- [Homepage](https://anaconda.org/)
	- [Using conda on the 'Omics lab server](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/using_conda.md) 
	- [conda user guide](https://docs.conda.io/projects/conda/en/latest/user-guide/index.html) 
- angsd
	- Add to PATH: `/scratch/Tools/angsd`
	- Add to PATH `/scratch/Tools/angsd/misc`
	- [Software page](http://www.popgen.dk/angsd/index.php/ANGSD)
	- [OMICs lab tutorial](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/angsd.md)
- Assign Taxonomy with BLAST
	- Add `/scratch/Tools/Assign-Taxonomy-with-BLAST` to PATH
	- [Homepage](https://github.com/Joseph7e/Assign-Taxonomy-with-BLAST) 
- beast
	- PATH `/scratch/Tools/beast/bin`
	- Beast has a wonderful [home page](http://www.beast2.org/) with lots of help options. It also has many, many [tutorials](https://www.beast2.org/tutorials/) available for it.
	- It is probably better to use th [CIPRES](https://www.phylo.org/) server.
- bedtools
	- static binary, must add `/scratch/Tools/` to your path
	- [Bedtools has good online documentation](https://bedtools.readthedocs.io/en/latest/)
- blast2bed
	- Add `/scratch/Tools/blast2bed` to your path
- blat
	- Add `/scratch/Tools/blat` to your path	 
- bwa
	- Add to PATH  `/scratch/Tools/bwa`
- diamond
	-  static binary, must add `/scratch/Tools/` to your path
	-  Use with the script [getRBH.pl](https://github.com/Computational-conSequences/SequenceTools), which was developed by [Hern치ndez-Salmer칩n & Moreno-Hagelsieb 2020](https://doi.org/10.1186/s12864-020-07132-6)
- docker
	- Talk to Sangeet about using this 
- Dsuite
	- Add to Path `/scratch/Tools/Dsuite/Build`
	- Add to Path `/scratch/Tools/Dsuite/utils/`
	- Must do `export LD_LIBRARY_PATH=/scratch/Tools/local/lib64/` before trying to run.
	- [Homepage](https://github.com/millanek/Dsuite)
	- [Online tutorial](https://github.com/millanek/tutorials/tree/master/analysis_of_introgression_with_snp_data)
- EMBOSS
	- The executables are in `/scratch/Tools/local/bin`
-  easySFS
- exonerate 2.2.0
	- Add to Path `/scratch/Tools/exonerate-2.2.0-x86_64/bin/`
- fastPHASE
	- static binary, must add `/scratch/Tools/` to your path
- FastQC
	- Add to Path `/scratch/Tools/FastQC/`
- FastTreeMP
	- static binary, must add `/scratch/Tools/` to your path
	- [Home page](http://www.microbesonline.org/fasttree/#How)
- gatk-4.2.0.0
	- Add to path `/scratch/Tools/gatk-4.2.0.0/'
	- [Home page](https://gatk.broadinstitute.org/hc/en-us)
- gcc 4.9.4
	- The executables are in `/scratch/Tools/local/bin` 
- genus_assigner
- ghostx-1.3.7
	- Add to path `/scratch/Tools/ghostx-1.3.7/src/`
- G-Phocs
- gsl-2.6
- htslib
- hal
-	- [Home page](https://github.com/ComparativeGenomicsToolkit/hal/blob/master/README.md)
- hmmer
	- Add to path `/scratch/Tools/hmmer-3.3.2/bin/`
	- [Wonderful manual](http://eddylab.org/software/hmmer/Userguide.pdf)
- maker
	- [Home page]()
	- [Online tutorial](https://github.com/Michigan-Mycology/Lab-Code-and-Hacks/tree/master/Genomics/Maker_pipeline) 
- mira
	- Add to path `/scratch/Tools/mira_4.0.2_linux-gnu_x86_64_static/bin/`
- MITObim
	- must add mira to your path for MITObim to work 
- mmseqs
	- Use with the script [getRBH.pl](https://github.com/Computational-conSequences/SequenceTools), which was developed by [Hern치ndez-Salmer칩n & Moreno-Hagelsieb 2020](https://doi.org/10.1186/s12864-020-07132-6)
- msmc-tools
- parallel
	- The executables are in `/scratch/Tools/local/bin`
	- Very, very handy tool that replaces for loops
	- [Helpful online manual](https://www.gnu.org/software/parallel/sphinx.html)
	- ['Omics lab tutorial]()
- picard.jar
- proj-8.1.0
- QuIBL
- RAiSD
- SequenceTools
	- Add to Path `/scratch/Tools/SequenceTools`
	- This is where [getRBH.pl](https://doi.org/10.1186/s12864-020-07132-6) lives
	- Useful for doing fast ortholog searches
- snapp_prep-master
	- Add to path `/scratch/Tools/snapp_prep-master`
	- [Online tutorial](https://github.com/elsemikk/tutorials/blob/master/divergence_time_estimation/README.md)
	- From [Stange et al. 2018](https://doi.org/10.1093/sysbio/syy006).
- Stacks 2.58
	- Must do `export LD_LIBRARY_PATH=/scratch/Tools/local/lib64/` before running
	- The executables are in `/scratch/Tools/local/bin`
- snpEff
	- [Home page](http://pcingola.github.io/SnpEff/)
	- ['Omics lab tutorial]() 
- treemix-1.13
	- Must do `export LD_LIBRARY_PATH=/scratch/Tools/local/lib/:/scratch/Tools/local/lib64/` before running
	- The executables are in `/scratch/Tools/local/bin`
- TrimGalore-0.6.6

## /scratch/Tools/ucsc_genome_browser_tools

All of the tools from the [University of California Santa Cruz Genome Browser](https://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/) are located here. This [site](https://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/) also contains the manual pages for each program.

- addCols    
- ameme      
- autoDtd    
- autoSql    
- autoXml    
- ave   
- aveCols    
- axtChain     
- axtSort    
- axtSwap    
- axtToMaf     
- axtToPsl     
- bamToPsl     
- barChartMaxLimit     
- bedClip    
- bedCommonRegions     
- bedCoverage  
- bedExtendRanges  
- bedGeneParts    
- bedGraphPack    
- bedGraphToBigWig     
- bedIntersect    
- bedItemOverlapCount 
- bedJoinTabOffset     
- bedJoinTabOffset.py     
- bedMergeAdjacent     
- bedPartition    
- bedPileUps      
- bedRemoveOverlap     
- bedRestrictToPositions     
- bedSort    
- bedToBigBed     
- bedToExons      
- bedToGenePred 
- bedToPsl     
- bedWeedOverlapping      
- bigBedInfo      
- bigBedNamedItems     
- bigBedSummary   
- bigBedToBed     
- bigGenePredToGenePred  
- bigHeat    
- bigMafToMaf     
- bigPslToPsl     
- bigWigAverageOverBed       
- bigWigCat       
- bigWigCluster   
- bigWigCorrelate      
- bigWigInfo      
- bigWigMerge     
- bigWigSummary   
- bigWigToBedGraph     
- bigWigToWig     
- binFromRange   
- blastToPsl      
- blastXmlToPsl   
- blat
- calc       
- catDir     
- catUncomment    
- chainAntiRepeat      
- chainBridge     
- chainCleaner    
- chainFilter     
- chainMergeSort       
- chainNet     
- chainPreNet     
- chainScore      
- chainSort       
- chainSplit      
- chainStitchId   
- chainSwap       
- chainToAxt      
- chainToPsl      
- chainToPslBasic      
- checkAgpAndFa   
- checkCoverageGaps  
- checkHgFindSpec      
- checkTableCoords   
- chopFaLines     
- chromGraphFromBin 
- chromGraphToBin       
- chromToUcsc     
- clusterGenes     
- clusterMatrixToBarChartBed  
- colTransform    
- countChars      
- cpg_lh      
- crTreeIndexBed       
- crTreeSearchBed      
- dbSnoop    
- dbTrash     
- endsInLf     
- estOrient        
- expMatrixToBarchartBed      
- faAlign    
- faCmp      
- faCount    
- faFilter     
- faFilterN       
- faFrag     
- faNoise    
- faOneRecord     
- faPolyASizes    
- faRandomize     
- faRc       
- faSize     
- faSomeRecords   
- faSplit    
- faToFastq       
- faToTab    
- faToTwoBit      
- faToVcf    
- faTrans    
- fastqStatsAndSubsample     
- fastqToFa       
- featureBits      
- fetchChromSizes      
- findMotif       
- fixStepToBedGraph.pl       
- gapToLift        
- genePredCheck    
- genePredFilter   
- genePredHisto    
- genePredSingleCover      
- genePredToBed    
- genePredToBigGenePred       
- genePredToFakePsl   
- genePredToGtf    
- genePredToMafFrames      
- genePredToProt   
- gensub2    
- getRna      
- getRnaPred       
- gff3ToGenePred   
- gff3ToPsl       
- gmtime     
- gtfToGenePred    
- headRest     
- hgBbiDbLink     
- hgFakeAgp       
- hgFindSpec       
- hgGcPercent      
- hgGoldGapGl      
- hgLoadBed        
- hgLoadChain      
- hgLoadGap        
- hgLoadMaf        
- hgLoadMafSummary      
- hgLoadNet        
- hgLoadOut        
- hgLoadOutJoined       
- hgLoadSqlTab    
- hgLoadWiggle     
- hgSpeciesRna     
- hgTrackDb        
- hgWiggle      
- hgsql      
- hgsqldump       
- hgvsToVcf        
- hicInfo     
- htmlCheck       
- hubCheck      
- hubClone      
- hubPublicCheck   
- ixIxx      
- lastz-1.04.00 
- lastz_D-1.04.00    
- lavToAxt     
- lavToPsl     
- ldHgGene      
- liftOver      
- liftOverMerge   
- liftUp      
- linesToRa       
- localtime        
- mafAddIRows     
- mafAddQRows     
- mafCoverage      
- mafFetch      
- mafFilter       
- mafFrag     
- mafFrags      
- mafGene     
- mafMeFirst      
- mafNoAlign      
- mafOrder     
- mafRanges       
- mafSpeciesList       
- mafSpeciesSubset     
- mafSplit     
- mafSplitPos      
- mafToAxt     
- mafToBigMaf     
- mafToPsl     
- mafToSnpBed      
- mafsInRegion    
- makeTableList    
- maskOutFa       
- matrixClusterColumns       
- matrixMarketToTsv       
- matrixNormalize      
- mktime     
- mrnaToGene       
- netChainSubset       
- netClass      
- netFilter       
- netSplit     
- netSyntenic     
- netToAxt     
- netToBed     
- newProg    
- newPythonProg   
- nibFrag    
- nibSize    
- oligoMatch      
- overlapSelect    
- para       
- paraFetch       
- paraHub    
- paraHubStop     
- paraNode     
- paraNodeStart   
- paraNodeStatus       
- paraNodeStop    
- paraSync     
- paraTestJob     
- parasol    
- positionalTblCheck       
- pslCDnaFilter    
- pslCat     
- pslCheck      
- pslDropOverlap       
- pslFilter       
- pslHisto     
- pslLiftSubrangeBlat      
- pslMap     
- pslMapPostChain      
- pslMrnaCover    
- pslPairs     
- pslPartition    
- pslPosTarget    
- pslPretty       
- pslRc      
- pslRecalcMatch       
- pslRemoveFrameShifts       
- pslReps    
- pslScore     
- pslSelect       
- pslSomeRecords       
- pslSort    
- pslSortAcc      
- pslStats     
- pslSwap    
- pslToBed     
- pslToBigPsl      
- pslToChain      
- pslToPslx       
- pslxToFa     
- qaToQac    
- qacAgpLift      
- qacToQa    
- qacToWig     
- raSqlQuery       
- raToLines       
- raToTab    
- randomLines     
- rmFaDups     
- rmskAlignToPsl       
- rowsToCols      
- sizeof     
- spacedToTab     
- splitFile       
- splitFileByColumn       
- sqlToXml     
- strexCalc       
- stringify       
- subChar    
- subColumn       
- tabQuery     
- tailLines       
- tdbQuery      
- tdbRename       
- tdbSort    
- textHistogram   
- tickToDate      
- toLower    
- toUpper    
- trackDbIndexBb   
- transMapPslToGenePred       
- trfBig     
- twoBitDup       
- twoBitInfo      
- twoBitMask      
- twoBitToFa      
- ucscApiClient   
- udr    
- vai.pl   
- validateFiles    
- validateManifest     
- varStepToBedGraph.pl       
- vcfToBed     
- webSync  
- wigCorrelate    
- wigEncode       
- wigToBigWig     
- wordLine     
- xmlCat     
- xmlToSql

## /scratch/Tools/scripts
- analyze_tree_asymmetry.rb
- blast_taxonomy_report.pl
- construct_unaln_files_180325.pl
- extract_blocks.rb
- fastaToTab
- fill_seq.rb
- getClassFasta.pl
- get_fixed_site_gts.rb
- get_number_of_pi_sites.rb
- get_number_of_variable_sites.rb
- get_parsimony_score.sh
- gff3ToFasta
- mask_imputed_gts.rb
- [parallelLastz.pl](https://github.com/jnarayan81/parallelLastz)
- plot_d.rb
- plot_f4ratio.rb
- plot_fixed_site_gts.rb
- plot_ggi.rb
- plotInvestigateResults.R
- plot_tree_asymmetry.rb
- print_tetramer_freqs_deg_filter_esom_VD.pl
- remove_seqs_if_all_gap.py
- summarize_ggi.sh
- tabToFasta
- vcf2phylip.py
	- From [here](https://github.com/edgardomortiz/vcf2phylip)
- remove_small_contigs.pl
	- From [here](https://github.com/elyons/BioInfoUtility)

# Misc.
- /opt/microsoft/omsagent/ruby/bin/ruby

# conda environments available on the server

To view a list of the conda environments available on the server, use `conda info --envs` or `conda env list`.

**NOTE:** After installing programs into a conda environment, one **MUST** run `chmod -R +777 /scratch/Tools/anaconda3` This ensures that other users can install a conda package after you.

- [admixture](https://dalexander.github.io/admixture/index.html)
- [agat](https://github.com/NBISweden/AGAT)
- [bcftools](https://samtools.github.io/bcftools/bcftools.html)
- [busco](https://busco.ezlab.org/)
- catadaptenv
- haltools
- metagenomics
	- includes cutaadapt 1.18
	- fastqc
	- hmmer
	- [humann2](https://huttenhower.sph.harvard.edu/humann2/)
	- [mafft](https://mafft.cbrc.jp/alignment/software/)
	- megahit
	- metaxa
- [py-popgen](https://pypi.org/project/py-popgen/)
- r_env
- rsem
- star
- [structure](https://web.stanford.edu/group/pritchardlab/structure.html)
  - includes [fastStructure](https://github.com/rajanil/fastStructure)
  - includes [structureHarvester](http://alumni.soe.ucsc.edu/~dearl/software/structureHarvester/)

- [raxml](https://github.com/stamatak/standard-RAxML)


# /home/sangeet/UserDirectories/Sangeet/Tools/

- admixture_linux-1.3.0
- AGAT
- App-cpanminus-1.5017
- bedtools
- bin
- bowtie2-2.4.1-linux-x86_64
- busco_downloads
- cgal-0.9.6-beta
- cmake-3.17.1-Linux-x86_64
- cmake-3.17.1-Linux-x86_64.sh
- core-rnammer
- fsc26_linux64
- gatk-4.1.3.0
- genometools-1.5.9
- hapflk1.1_linux64
- hapflk-1.4
- HMMER
- hmmer-3.3.2
- hyphy-2.5.11
- include
- install.sh
- javase-jdk13-downloads.html
- jellyfish-2.3.0
- kallisto
- kraken2-2.0.8-beta
- maker
- maker-functional
- MCScanX
- Miniconda
- myconfig.ini
- ncbi-blast-2.10.0+
	-['Omics lab tutorial](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/blast_quick_guide.md) 
- obsutil_linux_amd64_5.1.13
- OrthoFinder_2.3
- OrthoFinder-2.5.2
- PGDSpider_2.1.1.5
- plink-1.07-x86_64
- Reapr_1.0.18
- RepeatMasker
- RERconverge-0.3.0
- RNAMMER
- salmon-latest_linux_x86_64
- samtools-1.10
- satsuma2
- satsuma-code-0
- signalp-5.0b
- Spines_release_1.15
- sqlite-tools-linux-x86-3340100
- supernova-2.1.1
- tguenther-bayenv2_public-2b2b7f20bb62
- tmhmm-2.0c
- TransDecoder-TransDecoder-v5.5.0
- trinityrnaseq-v2.9.1
- v0.3.0.tar.gz
- vcftools_0.1.13


