Hmmer is a well-documented and useful tool for using sequence alignments to find homologous sequences in a target genome or genomes. I used it during my time at the University of Michigan to pull ~200 loci from ~46 genomes to infer a [well-resolved phylogeny of predacious and parasitic fungi in Zoopagales](https://doi.org/10.1016/j.ympev.2019.01.006). My scripts are available [here](https://github.com/wjdavis90/ZyGoLife/tree/master/phylogenomics) and are based on those developed [here](https://github.com/zygolife/Phylogenomics).

One of the things we are interested in for the *Philornis* project is whether or not the flies are beginning to show insecticide resistance. I very briefly ventured into this, and that is what I will use for the example.
 
According to [Scott 2016](https://doi.org/10.1002/ps.4328), there are three mutations of the voltage-sensitive sodium channel (Vssc) in *Musca domestica* that allow for resistance to pyrethroid insectides, primarily permethrin. These insecticdes work by prolonging the time the channel is open. The three mutations listed by Scott 2016 are: *kdr* (L1014F), *kdr-his* (L1014H), and *super-kdr* (M918T + L1014F). I figured this would be a good place to start to look for signs of insecticide resistance in *Philornis downsii*. 

To do so, I first downloaded full length sequences of Vssc of susceptible flies generated by [Sun et al. 2016](https://doi.org/10.1111/imb.12206) from GenBank. I then did a blastn search restricted to *M. domestica* and the sequences generated by [Kasai et al. 2017](https://doi.org/10.1111/imb.12267). I placed them into the same fasta file, cleaned up the names, and used [mafft online's](https://mafft.cbrc.jp/alignment/server/) `auto` strategy to align them.

Following a [stack overflow answer](https://www.biostars.org/p/132412/), I used [Easel](https://github.com/EddyRivasLab/easel) to convert the *M. domestica* Vssc alignment created above to [stockholm format](http://scikit-bio.org/docs/0.5.0/generated/skbio.io.format.stockholm.html).

```bash
esl-reformat stockholm Musca_dom_VSSC_mafft_algned.fa > Musca_dom_VSSC_mafft_algned.stockholm
```
Then I created a hmmer model.
 
```bash
hmmbuild Musca_dom_VSSC.hmm Musca_dom_VSSC_mafft_algned.stockholm
```
Next, I used a modified version of the [pipeline](https://github.com/Michigan-Mycology/Lab-Code-and-Hacks/tree/master/Phylogenomics) I used for ZygoLife to find the Vssc ortholog in the *Philornis* genome.
```bash
hmmsearch -E1e-8 --domtblout Pdown_Vssc.domtbl Musca_dom_VSSC.hmm Philornis_downsi1_v1.proteins.fasta &> Pdown_Vssc.log

perl get_best_hmmtbl.pl Pdown_Vssc.domtbl > Pdown_Vssc.best
```
