This is a repository for tutorials/examples I wrote for the Lamichhaney 'Omics Lab. There is also a [list](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/online_tutorials.md) of online tutorials that I have used and found helpful.

A common pipeline/work flow for the 'Omics lab is presented below. For all the examples, I assume three populations of 5 individuals each. I named the populations "August", "September", and "October", and the individuals are labelled as the first letter of the month (i.e., population name) followed by a number, e.g., A1.

1. [Align reads to a reference genome and call SNPS](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/Aligning_reads_calling_SNPS.md)
2. [Calculate intra-population statistics](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/intra_population_stats.md)
3. [Determine relatedness of individuals within populations](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/relatedness.md)
4. [Use Treemix to look at migration among populations](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/treemix.md)
5. Estimate divergence among populations using either [Fst](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/fst.md) or [Population branch statistics](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/angsd.md)
6. Identify windows of interest
7. [Build phylogenetic trees of the windows of interest](https://github.com/wjdavis90/Omics_lab_server/blob/main/tutorials/vcf_to_fast_tree.md)

